server/src/main/java/com/flipkart/varadhi/services/DlqService.java

```diff
@@ -34,37 +42,55 @@ public CompletableFuture<SubscriptionOperation> unsideline(
-                            return consumer.getMessages(messagesRequest).whenComplete((messages, t) -> {
-                                // write partial response i.e. return messages from single shard or error.
-                                GetMessagesResponse shardResponse = GetMessagesResponse.of(assignment.getShardId());
-                                if (t == null) {
-                                    shardResponse.setMessages(messages);
-                                } else {
-                                    log.error(
-                                            "Failed to getMessages for {}(shard:{}). Error:{}", subscription.getName(),
-                                            assignment.getShardId(), t.getMessage()
-                                    );
-                                    shardResponse.setError(t.getMessage());
-                                }
-                                log.info("Sending a GetMessage response for {}", assignment);
-                                responseWriter.accept(shardResponse);
-                            });
-                        }).toArray(CompletableFuture[]::new)));
+        ShardDlqMsgResponseCollector finalResponse = new ShardDlqMsgResponseCollector();
+        boolean isRequestByTimeStamp = earliestFailedAt != UNSPECIFIED_TS;
+        return controllerClient.getShardAssignments(subscription.getName()).thenCompose(assignments -> {
+            List<CompletableFuture<ShardDlqMessageResponse>> shardFutures = new ArrayList<>();
+            for (Assignment a : assignments.getAssignments()) {
+                int shardId = a.getShardId();
+                if (!isRequestByTimeStamp && !pageMarkers.hasMarker(shardId)) {
+                    log.info("Shard {} has no markers, skipping getMessages().", shardId);
+                    continue;
+                }
+                shardFutures.add(getMessagesForShard(isRequestByTimeStamp, a.getConsumerId(), earliestFailedAt,
+                        pageMarkers.getShardMarker(shardId), limit
+                ).whenComplete((r, t) -> processShardResponse(shardId, responseWriter, r, t, finalResponse)));
+            }
+            return CompletableFuture.allOf(shardFutures.toArray(new CompletableFuture[0]));
+        }).whenComplete((v, t) -> responseWriter.accept(finalResponse.toAggregatedResponse()));
+    }
+
+    private void processShardResponse(
+            int shardId, Consumer<DlqMessagesResponse> responseWriter, ShardDlqMessageResponse r, Throwable t,
+            ShardDlqMsgResponseCollector finalResponse
+    ) {
+        if (r != null && !r.getMessages().isEmpty()) {
+            log.info(
+                    "shard {} returned {} messages nextMarker {}.", shardId, r.getMessages().size(),
+                    r.getNextPageMarker()
+            );
+            responseWriter.accept(DlqMessagesResponse.of(r.getMessages()));
+        }
+        finalResponse.collectShardResponse(shardId, t, r);
```

Comment:
so it looks like, in the response, we are writing an array?
having "shard + 1" elements. where the first "shard" elements are the messages per shard and the last element is the info about the pageMarkers and errors.

Is this understanding correct?

Reply:
yes, 1 response per shard and followed by final response with all marker/errors.

Reply:
Around the chunked response. The overall repsonse is  not a valid json
so that is an issue.

I am expecting that overall, the response headers will contain `content-type : application/json`.
which will cause the query on UI to break, as that will try to parse the chunked response as a full json